modules/apps/portal-remote/portal-remote-cors-impl/src/main/java/com/liferay/portal/remote/cors/internal/DynamicURLToCORSSupportMapper.java

```diff
@@ -0,0 +1,269 @@
+
+		if (urlPath.charAt(0) != '/') {
+			onlyExact = true;
+		}
+		else if ((urlPath.length() > 1) &&
+				 (urlPath.charAt(urlPath.length() - 2) == '/') &&
+				 (urlPath.charAt(urlPath.length() - 1) == '*')) {
+
+			onlyWildcard = true;
+		}
+
+		CORSSupport corsSupport = null;
+
+		TrieNode currentTrieNode = null;
+		TrieNode previousTrieNode = _wildCardTrieNode;
+
+		for (int i = 0; i < urlPath.length(); ++i) {
+			currentTrieNode = previousTrieNode.next(urlPath.charAt(i));
+
+			if (currentTrieNode == null) {
+				break;
+			}
+
+			if (!onlyExact && (urlPath.charAt(i) == '/')) {
+				TrieNode nextTrieNode = currentTrieNode.next('*');
+
+				if ((nextTrieNode != null) && nextTrieNode.isEnd()) {
+					corsSupport = nextTrieNode.get();
+				}
+			}
+
+			previousTrieNode = currentTrieNode;
+		}
+
+		if (currentTrieNode != null) {
+			if (onlyExact) {
+				if (!currentTrieNode.isEnd()) {
+					return null;
+				}
+
+				return currentTrieNode.get();
+			}
+
+			if (!onlyWildcard && currentTrieNode.isEnd()) {
+				return currentTrieNode.get();
+			}
+
+			currentTrieNode = currentTrieNode.next('/');
+
+			if (currentTrieNode != null) {
+				currentTrieNode = currentTrieNode.next('*');
+
+				if ((currentTrieNode != null) && currentTrieNode.isEnd()) {
+					corsSupport = currentTrieNode.get();
+				}
+			}
+		}
+
+		return corsSupport;
+	}
+
+	@Override
+	protected void put(
+		CORSSupport corsSupport, String urlPattern, boolean wildcard) {
+
+		TrieNode previousTrieNode = null;
+
+		if (wildcard) {
+			previousTrieNode = _wildCardTrieNode;
+		}
+		else {
+			previousTrieNode = _extensionTrieNode;
+		}
+
+		TrieNode currentTrieNode = null;
+
+		for (int i = 0; i < urlPattern.length(); ++i) {
+			int index = i;
+
+			if (!wildcard) {
+				index = urlPattern.length() - 1 - i;
+			}
+
+			currentTrieNode = previousTrieNode.next(urlPattern.charAt(index));
+
+			if (currentTrieNode == null) {
+				TrieNode nextNode = _trieNodeHeap.nextAvailableNode();
+
+				currentTrieNode = previousTrieNode.setNext(
+					urlPattern.charAt(index), nextNode);
+			}
+
+			previousTrieNode = currentTrieNode;
+		}
+
+		if (currentTrieNode != null) {
+			currentTrieNode.set(corsSupport);
+		}
+	}
+
+	private final TrieNode _extensionTrieNode;
+	private final TrieNodeHeap _trieNodeHeap;
+	private final TrieNode _wildCardTrieNode;
+
+	private static class TrieNode {
+
+		public TrieNode() {
+			_trieNodes = new ArrayList<>(ASCII_CHARACTER_RANGE);
+
+			for (int i = 0; i < ASCII_CHARACTER_RANGE; ++i) {
+				_trieNodes.add(null);
+			}
+		}
+
+		public CORSSupport get() {
+			return _corsSupport;
+		}
+
+		public boolean isEnd() {
+			if (_corsSupport != null) {
+				return true;
+			}
+
+			return false;
+		}
+
+		public TrieNode next(char character) {
+			int index = character - ASCII_PRINTABLE_OFFSET;
+
+			if ((index < 0) ||
+				(index >= (ASCII_CHARACTER_RANGE + ASCII_PRINTABLE_OFFSET))) {
+
+				throw new IllegalArgumentException();
+			}
```

Comment:
I think we should drop this and let the outer catch block handle this and turn it into the `IllegalArgumentException`

Reply:
once the block is removed the `index` variable can be inlined

Reply:
Right, will do.